### PR TITLE
XML-based pipeline and ligand parametrization

### PR DESCRIPTION
**Pull request by bas-rustenburg**
This PR adds a `ForceFieldProtonDrive` class that uses a regular openmm `Topology`, and an improved protons format FFXML file. Also included is a parametrization pipeline that runs a given small molecule through Epik and then uses openmoltools to generate forcefield templates, which are then compiled into the new format.

Here is an example of the new format, for imidazole with 2 protonation states.

https://gist.github.com/bas-rustenburg/1b824747b4f0b545c6992a23e7c2e688

Things to address in separate pull-request:

* Update calibration procedure to work with new driver [next top priority]
* Update docs to describe new set up procedure, and ffxml format [in the new year]

### CODE REVIEW

**Comment by bas-rustenburg** [comment]
Proceeding to merge `master` now that @gregoryross has finished his PR in #63 .

**Comment by bas-rustenburg** [comment]
@gregoryross : Could you review? Especially pay attention to whether I might have accidentally broken your ncmc implementation in the merge, and whether I added it to the ForceFieldProtonDrive correctly.

File: protons/driver.py [hunk 1]
```diff
@@ -974,7 +1003,7 @@ def _attempt_state_change(self, context, reject_on_nan=False):
                     self.nrejected += 1
                 self.pot_energies.append(pot1)
                 self.kin_energies.append(kin1)
-                # Restore sams_sampler states.
+                # Restore titration states.
                 for titration_group_index in titration_group_indices:
                     self._set_titration_state(titration_group_index, initial_titration_states[titration_group_index], context)
                 # If using NCMC, restore coordinates and flip velocities.
```
Comment (gregoryross) [inline protons/driver.py]: The `_ncmc_ghmc` and `_attempt_state_change` functions are unaffected, which is good for us.

File: protons/driver.py [hunk 2]
```diff
@@ -1522,6 +1549,315 @@ def __init__(self, system, temperature, pH, prmtop, cpin_filename, integrator, p
...
+            self._set_titration_state(group_index, 0)
+
+        # If number of attempts not specified, number of attempts is set equal to the number of titratable groups
+        self._set_num_attempts_per_update(nattempts_per_update)
+
+        # Reset statistics.
+        # nattempted = 0
+        # naccepted = 0
+        # nrejected = 0
+        # work_history = list()
```
Comment (gregoryross) [inline protons/driver.py]: These seem safe to delete.

File: protons/ligands.py [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+
+        Returns
+        -------
+        bool - True if the atom is part of the bond
+        """
+        return atomname in (self.atomName1, self.atomName2)
+
+
+class _State(object):
+    """Private class representing a template of a single isomeric state of the molecule.
```
Comment (gregoryross) [inline protons/ligands.py]: format

File: protons/ligands.py [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+        for name, atom in self.atoms.items():
+            if atom is None:
+                dummies.append(name)
+            elif atom.is_dummy():
+                dummies.append(name)
+
+        return dummies
+
+    def set_atom(self, atom):
+        """Set the parameters for a single atom
```
Comment (gregoryross) [inline protons/ligands.py]: Format code paragraph

File: protons/ligands.py [hunk 5]
```diff
@@ -0,0 +1,874 @@
...
+        Parameters
+        ----------
+        bonded_to - dictionary of the atoms its bonded to, and their potential types
+        params - gaff parameters for all atomtypes
+        type_list - all possible types for this atom
+
+        Returns
+        -------
+
+        solutions : a
```
Comment (gregoryross) [inline protons/ligands.py]: Formating

File: protons/ligands.py [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+
+            if have_found_valid_type:
+                # Store the information on the potential atom types of the bond partners
+                # This can help narrow down which types these atoms can take on in the final solution
+                solutions[atom_type] = valid_match
+
+        return solutions
+
+    def _retrieve_atom_type_parameters(self, atom_type_name):
+        """ Look through FFXML files and find all parameters pertaining to the supplied atom type.
```
Comment (gregoryross) [inline protons/ligands.py]: format

**Comment by bas-rustenburg** [comment]
Need to switch the builds to openmm 7.1 dev.

**Pull request merged**